src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -6991,6 +7013,60 @@ static bool dependsOnTypeParameter(TypeParameterSymbol typeParameter1, TypeParam
+
+                        var consequence = conditional.Consequence;
+                        (BoundExpression consequenceOperand, Conversion consequenceConversion) = RemoveConversion(consequence, includeExplicitConversions: false);
+                        var consequenceRValue = ConvertConditionalOperandOrSwitchExpressionArmResult(consequence, consequenceOperand, consequenceConversion, targetTypeWithNullability, info[0].ResultType, info[0].State, info[0].EndReachable);
+                        var alternative = conditional.Alternative;
+                        (BoundExpression alternativeOperand, Conversion alternativeConversion) = RemoveConversion(alternative, includeExplicitConversions: false);
+                        var alternativeRValue = ConvertConditionalOperandOrSwitchExpressionArmResult(alternative, alternativeOperand, alternativeConversion, targetTypeWithNullability, info[1].ResultType, info[1].State, info[1].EndReachable);
+
+                        _conditionalInfoForConversionOpt.Remove(conditional);
+
+                        return consequenceRValue.State.Join(alternativeRValue.State);
+
+                    case BoundConvertedSwitchExpression { WasTargetTyped: true } @switch:
+                        Debug.Assert(_conditionalInfoForConversionOpt is not null && _conditionalInfoForConversionOpt.ContainsKey(@switch));
+                        info = _conditionalInfoForConversionOpt[@switch];
+
+                        Debug.Assert(info.Length == @switch.SwitchArms.Length);
+
+                        var resultTypes = ArrayBuilder<TypeWithState>.GetInstance(info.Length);
+
+                        for (int i = 0; i < info.Length; i++)
+                        {
+                            var (state, armResultType, isReachable) = info[i];
+                            var arm = @switch.SwitchArms[i].Value;
+                            var (operand, conversion) = RemoveConversion(arm, includeExplicitConversions: false);
+                            resultTypes.Add(ConvertConditionalOperandOrSwitchExpressionArmResult(arm, operand, conversion, targetTypeWithNullability, armResultType, state, isReachable));
+                        }
+
+                        var resultState = BestTypeInferrer.GetNullableState(resultTypes);
+                        resultTypes.Free();
+                        _conditionalInfoForConversionOpt.Remove(@switch);
+                        return resultState;
+
+                    case BoundObjectCreationExpression { WasTargetTyped: true }:
+                    case BoundUnconvertedObjectCreationExpression:
+                        return NullableFlowState.NotNull;
+
+                    case BoundUnconvertedConditionalOperator:
+                    case BoundUnconvertedSwitchExpression:
+                        return operandType.State;
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(conversionOperand.Kind);
+                }
+            }
```

Comment:
Pausing here for the afternoon. I think I get the necessity of this change but I don't quite grasp yet what makes the dictionary necessary as opposed to keeping the additional states on the stack. Maybe we can go over it on Monday.